### PR TITLE
std::fmt: Various fixes and improvements to documentation

### LINKED ISSUE #71088: Format string parser does not fail on invalid input
This code:

```rust
struct Foo;

impl Display for Foo {
  fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
    write!(f, "{}", f.fill() as u32)
  }
}

fn main() {
  println!("{:\t}", Foo)
}
```

Compiles, and prints `32`. According to the `std::fmt` docs, 

```
[[fill]align][sign]['#']['0'][width]['.' precision][type]
```

is the syntax of format specifications; however, `\t` does not fulfill any of these elements. I assumed that it might accidentally be implementing the syntax

```
[fill][align][sign]['#']['0'][width]['.' precision][type]
```

however, if that were true, then this should print `9` (and not `32`). I haven't looked into the parsing code, but will after filing this issue.

### LINKED ISSUE #96413: std::fmt: Precision via asterisk does not fit documentation when using named or explicit positional arguments
From the documentation on `std::fmt`, section "Precision" (highlighting in bold by me):

> There are three possible ways to specify the desired precision:
> 
> [...]
> 
> 3. An asterisk `.*`:
> 
>    `.*` means that this `{...}` is associated with two* format inputs rather than one: the first input holds the `usize` precision, and the second holds the value to print. **Note that in this case, if one uses the format string `{<arg>:<spec>.*}`, then the `<arg>` part refers to the *value* to print, and the `precision` must come in the input preceding `<arg>`.**

The bold part does not fit the current behavior. Consider the following snippet:
```
fn main() {
    println!("{2:.*} {}", "foobar", 1, 2.0);
}
```
[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3ddd7a47216da3385d0476ee4b5a60bb)

According to the documentation, one would expect this program to output `2.0 foobar`. Instead, rustc fails to compile it:
```
error[[E0308]](https://doc.rust-lang.org/nightly/error-index.html#E0308): mismatched types
   [--> src/main.rs:2:26
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)    |
2   |     println!("{2:.*}{}", "foobar", 1, 2.0);
    |     ---------------------^^^^^^^^---------
    |     |                    |
    |     |                    expected `usize`, found `&str`
    |     arguments to this function are incorrect
    |
    = note: expected reference `&usize`
               found reference `&&str`
note: associated function defined here
    = note: this error originates in the macro `$crate::format_args_nl` (in Nightly builds, run with -Z macro-backtrace for more info)
```

After glancing at the implementation, it looks to me like in the asterisk case, the precision is always the next implicit positional argument: https://github.com/rust-lang/rust/blob/055bf4ccd521c2c2185166c86951be7be145727c/compiler/rustc_parse_format/src/lib.rs#L577-L585

The current implementation determines the argument that will be formatted only after the argument containing the precision. This is the reason why the documentation fits the behavior for implicit positional parameters (like `{:.*}`). However, this doesn't help in the case of explicit positional parameters (like `{42:.*}`) and named parameters (like `{foobar:.*}`) – the `precision` usually does not come in the input preceding `<arg>`, as stated by the documentation.

In my opinion, this is not an implementation bug, but a documentation bug – the behavior hasn't changed since at least Rust 1.0, see https://godbolt.org/z/xf61oP6hE. It would probably be unwise to change it now as some crates may depend on it. I'd be happy to open a pull request fixing the documentation – there are a few other things I want to enhance in the `std::fmt` documentation anyway.

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (18f314e70 2022-04-24)
binary: rustc
commit-hash: 18f314e7027fe7084aaab8620c624a0d7bd29e70
commit-date: 2022-04-24
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.1
```

### PR DESCRIPTION
**Pull request by EliasHolzmann**
This PR contains the following changes:

- **Added argument index comments to examples for specifying precision**
  
  The examples for specifying the precision have comments explaining which
  argument the specifier is referring to. However, for implicit positional
  arguments, the examples simply refer to "next arg". To simplify following the
  comments, "next arg" was supplemented with the actual resulting argument index.
  
- **Fixed documentation for specifying precision via `.*`**
  
  The documentation stated that in case of the syntax `{<arg>:<spec>.*}`, "the
  `<arg>` part refers to the value to print, and the precision must come in the
  input preceding `<arg>`". This is not correct: the <arg> part does indeed refer
  to the value to print, but the precision does not come in the input preciding
  arg, but in the next implicit input (as if specified with {}).
  
  Fixes #96413.
  
  
  
- **Fix the grammar documentation**
  
  According to the grammar documented, the format specifier `{: }` should not be
  legal because of the whitespace it contains. However, in reality, this is
  perfectly fine because the actual implementation allows spaces before the
  closing brace. Fixes #71088.
  
  Also, the exact meaning of most of the terminal symbols was not specified, for
  example the meaning of `identifier`.
  
- **Removed reference to Formatter::buf and other private fields**
  
  Formatter::buf is not a public field and therefore isn't very helpful in user-
  facing documentation. Also, the other public fields of Formatter were removed
  during stabilization of std::fmt (4af3494bb0) and can only be accessed via
  getters.
  
- **Improved list of formatting macros**
  
  Two improvements:
  1. write! can not only receive a `io::Write`, but also a `fmt::Write` as first argument.
  2. The description texts now contain links to the actual macros for easier
     navigation.

### CODE REVIEW

File: library/alloc/src/fmt.rs [hunk 1]
```diff
@@ -238,15 +240,15 @@
...
+//! // Hello {next arg -> arg 0 ("x")} is {second of next two args -> arg 2 (0.01) with precision
+//! //                          specified in first of next two args -> arg 1 (5)}
 //! println!("Hello {} is {:.*}",    "x", 5, 0.01);
 //!
-//! // Hello {next arg ("x")} is {arg 2 (0.01) with precision
-//! //                          specified in its predecessor (5)}
-//! println!("Hello {} is {2:.*}",   "x", 5, 0.01);
+//! // Hello {arg 1 ("x")} is {arg 2 (0.01) with precision
+//! //                          specified in next arg -> arg 0 (5)}
+//! println!("Hello {1} is {2:.*}",  5, "x", 0.01);
```
Comment (joshtriplett) [inline library/alloc/src/fmt.rs:249]: I think both of these examples are important. The second shows that the argument and precision can be discontiguous; the first shows that "next argument" takes previous positional arguments into account.

**Comment by joshtriplett** [comment]
Posted one minor nit; LGTM otherwise.

File: library/alloc/src/fmt.rs (hunk 1 above)
Comment (EliasHolzmann) [reply library/alloc/src/fmt.rs]: Thanks; I agree. Fixed now. (Also, I realized that the first commit contained some changes that belonged to the second and vice versa – must have happened during interactive rebase, fixed that, too).

**Pull request merged**